From abac8566212b39d58a519240f2af2c374947e312 Mon Sep 17 00:00:00 2001
From: Asen Alexandrov <alexandrov@vmware.com>
Date: Thu, 30 Mar 2023 09:29:55 +0300
Subject: [PATCH 10/15] chore: Add more logs in wasi_socket_ext

 100.0% wasmedge_stubs/

 100.0% wasmedge_stubs/
diff --git a/wasmedge_stubs/wasi_socket_ext.c b/wasmedge_stubs/wasi_socket_ext.c
index 0c74c75a..5a2a3d26 100644
--- a/wasmedge_stubs/wasi_socket_ext.c
+++ b/wasmedge_stubs/wasi_socket_ext.c
@@ -184,6 +184,7 @@ int32_t __imported_wasmedge_wasi_snapshot_preview1_sock_setsockopt(
 
 int socket(int domain, int type, int protocol)
 {
+  WSEDEBUG("WWSock| socket called: %d, %d, %d \n", domain, type, protocol);
   int fd;
   address_family_t af = (domain == AF_INET ? kInet4 : kInet6);
   socket_type_t st = (type == SOCK_STREAM ? kStream : kDatagram);
@@ -191,6 +192,7 @@ int socket(int domain, int type, int protocol)
   if (0 != res) {
     errno = res;
     printf("socket err: %s \n", strerror(errno));
+    WSEDEBUG("WWSock| socket failed with error: %s \n", strerror(errno));
     return -1;
   }
   WSEDEBUG("WWSock| socket returning: %d \n", fd);
@@ -271,8 +273,10 @@ int accept(int fd, struct sockaddr *restrict addr, socklen_t *restrict len) {
       fd, (uint32_t *)&new_sockfd);
   if (res != 0) {
     errno = res;
+    WSEDEBUG("WWSock| accept[%d]: failed=%d\n", __LINE__, errno);
     return -1;
   }
+  WSEDEBUG("WWSock| accept[%d]: client_fd=%d\n", __LINE__, new_sockfd);
   return new_sockfd;
 }
 
-- 
2.38.1

