From 547ab191f123a4a7cf157bf0e20a5eb4ae4222b9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jesu=CC=81s=20Gonza=CC=81lez?= <jesusgm@vmware.com>
Date: Tue, 14 Feb 2023 19:09:34 +0100
Subject: [PATCH 06/18] fix: Avoid `#include <syslog.h>` since it is not part
 of wasi-libc

 100.0% ext/gd/libgd/

 100.0% ext/gd/libgd/
diff --git a/ext/gd/libgd/gd_errors.h b/ext/gd/libgd/gd_errors.h
index 8a28bf60519..e84cef4578b 100644
--- a/ext/gd/libgd/gd_errors.h
+++ b/ext/gd/libgd/gd_errors.h
@@ -1,11 +1,13 @@
 #ifndef GD_ERRORS_H
 #define GD_ERRORS_H
 
-#ifndef _WIN32
-# include <syslog.h>
-#else
-# include "win32/syslog.h"
-#endif
+#ifndef __wasi__
+    #ifndef _WIN32
+    # include <syslog.h>
+    #else
+    # include "win32/syslog.h"
+    #endif
+#endif /* __wasi__ */
 
 /*
 LOG_EMERG      system is unusable
-- 
2.38.1

